### PR TITLE
logs: Add durable identifier log record attributes

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Defines `log.record.id` and `log.record.name` which may be added when Log Records have the concept of durable identification

## Merge requirement checklist

* [ ] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/attributes-registry/log.md [hunk 1]
```diff
@@ -40,12 +40,68 @@ Attributes for a file to which log was emitted.
...
+| `log.record.original` | string | The complete orignal Log Record. [3]          | `77 <86>1 2015-08-06T21:58:59.694Z 192.168.2.133 inactive - - - Something happened`; `[INFO] 8/3/24 12:34:56 Something happened` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `log.record.uid`      | string | A unique identifier for the Log Record. [4]   | `01ARZ3NDEKTSV4RRFFQ69G5FAV`                                                                                                     | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+
+**[1]:** This value MAY be added when processing a Log Record which has the
+concept of a durable identifier.
+
+A durable identifier is a unique value assigned to a well-known Log Record
+inside a given instrumentation scope. A durable identifier SHOULD be traceable
+back to a specific piece of code where the log is defined. Durable identifiers
+MAY be used by backends to safely and efficiently filter Log Records which
```
Comment (cijothomas) [inline docs/attributes-registry/log.md:56]:
```suggestion
MAY be used to safely and efficiently filter Log Records which
```

filtering could occur much before backends are reached....

**Review by jsuereth — CHANGES_REQUESTED** [review]
These fields are not defined in our attribute registry yet.  You'll need to first move these there and then add new fields.

This means these need to be defined in the `model/` YAML files.  See: https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#1-modify-the-yaml-model

Sorry - didn't realize we hadn't gotten these moved over yet.

**Comment by trisch-me** [comment]
@jsuereth we do have it moved and this file is autogenerated. 
https://github.com/open-telemetry/semantic-conventions/blob/main/model/registry/log.yaml

@CodeBlanch please update above mentioned yaml file in order to reflect changes in md. You can run `make attribute-registry-generation` to regenerate it

File: docs/attributes-registry/log.md [hunk 2]
```diff
@@ -40,12 +40,68 @@ Attributes for a file to which log was emitted.
...
+Two log helpers are defined inside the `MyLogic` instrumentation scope:
+`LogFoodPriceChanged` (with the durable identifier `1`) and
+`LogFoodRecallNotice` (with the durable identifier `2`). All Log Records emitted
+using `LogFoodPriceChanged` will have `log.record.id=1` and all Log Records
+emitted using `LogFoodRecallNotice` will have `log.record.id=2`.
+
+**[2]:** This value MAY be added when processing a Log Record which has the
+concept of a durable name.
+
+A durable name is similar to a durable identifier except it is more friendly
```
Comment (cijothomas) [inline docs/attributes-registry/log.md:98]:
```suggestion
A durable name is the human readable name for the durable identifier.
```

File: docs/attributes-registry/log.md [hunk 3]
```diff
@@ -40,12 +40,68 @@ Attributes for a file to which log was emitted.
...
-The id MAY be an [Universally Unique Lexicographically Sortable Identifier (ULID)](https://github.com/ulid/spec), but other identifiers (e.g. UUID) may be used as needed.
+| Attribute             | Type   | Description                                   | Examples                                                                                                                         | Stability                                                        |
+| --------------------- | ------ | --------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------- |
+| `log.record.id`       | string | A durable identifier for the Log Record. [1]  | `1`; `0x100F`                                                                                                                    | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (cijothomas) [inline docs/attributes-registry/log.md:45]:
```suggestion
| `log.record.id`       | int| A durable identifier for the Log Record. [1]  | `1`; `0x100F`                                                                                                                    | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

Unless there is a strong reason, consider making the durable id as an integer number, to make fast filtering etc. in hot path.

**Comment by CodeBlanch** [comment]
@trisch-me @jsuereth Updated the model. Haven't added CHANGELOG entry yet though, waiting to see if this moves forward first.

File: docs/attributes-registry/log.md [hunk 4]
```diff
@@ -40,12 +40,71 @@ Attributes for a file to which log was emitted.
...
+concept of a durable identifier.
+
+A durable identifier is a unique value assigned to a well-known Log
+Record inside a given instrumentation scope. A durable identifier
+SHOULD be traceable back to a specific piece of code where the log is
+defined. Durable identifiers MAY be used to safely and efficiently
+filter Log Records which produce high volume and yield low value
+(spammy logs).
+
+Consider this pseudo-code example:
```
Comment (cijothomas) [inline docs/attributes-registry/log.md:60]: this is actually csharp code, did you intend to switch to pseudo code?
logger.log(eventid, eventname, rest-of-logging-things-like-body-attributes-severity-etc)

File: docs/attributes-registry/log.md [hunk 5]
```diff
@@ -40,12 +40,71 @@ Attributes for a file to which log was emitted.
...
-The id MAY be an [Universally Unique Lexicographically Sortable Identifier (ULID)](https://github.com/ulid/spec), but other identifiers (e.g. UUID) may be used as needed.
+| Attribute             | Type   | Description                                  | Examples                                                                                                                         | Stability                                                        |
+| --------------------- | ------ | -------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------- |
+| `log.record.id`       | int    | A durable identifier for the Log Record. [1] | `1`; `10018`                                                                                                                     | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (tigrannajaryan) [inline docs/attributes-registry/log.md:45]: Are there any uniqueness requirements/guarantees for the id? Do we expect it to be globally unique? Unique within some limited scope? Are there any mechanisms for preventing id conflicts?

We also have `code.*` attributes that can point to a particular place in the code that emits the log record that can serve as sort of durable identifier (e.g. filename+function+line). What's the interplay (if any) between these and the newly proposed attribute?

File: docs/attributes-registry/log.md [hunk 6]
```diff
@@ -40,12 +40,71 @@ Attributes for a file to which log was emitted.
...
+| `log.record.name`     | string | A durable name for the Log Record. [2]       | `RequestProcessed`; `InvalidResponse`                                                                                            | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (tigrannajaryan) [inline docs/attributes-registry/log.md:46]: How are `id` and `name` related? Is `name` supposed to be the human-readable equivalent of `id`? Can different `id`-ed logrecords have the same `name`?

**Review by tigrannajaryan — COMMENTED** [review]
Has this proposal been discussed previously? If yes are there any recordings of previous discussions that can be linked to?

**Review by jsuereth — COMMENTED** [review]
Thanks for fixing the YAML portions.

I think `log.record.name` is problematic compared to `event.name` as folks may have added.  I'll let others closer to log/event signals discuss details, but from a general-semconv standpoint, I don't see issues in YAML/markdown setup.

File: docs/attributes-registry/log.md (hunk 6 above)
Comment (cijothomas) [reply docs/attributes-registry/log.md:46]: > Is name supposed to be the human-readable equivalent of id

Yes.

> Can different id-ed logrecords have the same name?

Not in well behaved application.

File: docs/attributes-registry/log.md (hunk 5 above)
Comment (cijothomas) [reply docs/attributes-registry/log.md:45]: filename/function/line etc. is not really durable, as it can change just with code refactoring!

Here is an example of EventId/Name: https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#log-event-id

Comment (tigrannajaryan) [reply docs/attributes-registry/log.md:45]: OK, makes sense regarding code references.

I think we still need clarity around uniqueness, conflicts, etc.

File: docs/attributes-registry/log.md (hunk 6 above)
Comment (lmolkova) [reply docs/attributes-registry/log.md:46]: Based on today discussion in the Event SIG, I believe we should use `event.name` for the durable event name. 

The TL;DR of the discussion

If logging API allows to define event name (like .NET, C++ and maybe Rust do), e.g.

```csharp
logger.LogInformation(new EventId(id: 42, name: "com.foo.my-event-name"), "something important")
```

then the corresponding name should be recorded as an `event.name`.

It's a user responsibility to have a unique name, there is nothing we can do to enforce it if users call Event API directly. 

So if user does `logger.LogInformation(new EventId(42, name: "foo"), "something important")` the best that logging bridge can do is to use the name that user provided and set `event.name = foo`. 

I wonder if other participants share the same understanding @CodeBlanch @MSNev @trask @breedx-splk

Comment (tigrannajaryan) [reply docs/attributes-registry/log.md:46]: Related to this, [early versions of Log data model](https://github.com/open-telemetry/oteps/blob/main/text/logs/0097-log-data-model.md#field-name) had a `Name` field defined as:

>Short event identifier that does not contain varying parts. Name describes what happened (e.g. "ProcessStarted"). Recommended to be no longer than 50 characters. Not guaranteed to be unique in any way. Typically used for filtering and grouping purposes in backends. This field is optional.

This seems pretty close to what the proposed `log.record.name` is and also to existing `event.name`. It almost feels like we should have help the `Name` field on LogRecord. I [commented here](https://github.com/open-telemetry/opentelemetry-specification/issues/3406#issuecomment-2307268466) that I think we need consider adding it back and then it would replace `event.name` and the proposed `log.record.name`.

Comment (MSNev) [reply docs/attributes-registry/log.md:46]: > Based on today's discussion in the Event SIG, I believe we should use `event.name` for the durable log name.
> 
> The gist of the discussion:
> 
> If logging API allows to define event name (like .NET, C++ and maybe Rust do), e.g.
> 
> ```cs
> logger.LogInformation(new EventId(id: 42, name: "com.foo.my-event-name"), "something important")
> ```
> 
> then the corresponding name should be recorded as an `event.name`.
> 
> It's a user responsibility to provide a unique name, there is nothing we can do to enforce it if users call Event API directly.
> 
> So if user does `logger.LogInformation(new EventId(42, "foo"), "something important")` the best that logging bridge can do is to use the name that user provided and set `event.name = foo`.
> 
> I wonder if other participants share the same understanding @CodeBlanch @MSNev @trask @breedx-splk

This is effectively what we talked about, however, as mentioned in the call my point of view (and recommendation) is still that otel "should" define that for this scenario there is a generic prefix that is prepended to `event.name` to ensure that these types of `events` cannot directly clash with any other OTel defined event.

During the meeting I proposed `log.*` but perhaps better options would be either `app.*`, `user.*` or `custom.*` so that it is explicitly called for backend processors of "known" event names that this is a custom event and therefore should be treated as such, as opposed to requiring them to enforce (more explicit) validity checks on any possible "known" events than necessary).

This should also NOT preclude the ability of languages reusing their existing logging mechanism's to "generate" true valid OTel defined events (without any prefix automatically added). Which could be done (potentially) by the introduction of some other explicitly named "element" (class, provider, property, attribute etc)

Comment (lmolkova) [reply docs/attributes-registry/log.md:46]: is the goal of the prefix to prevent users from accidentally using a name that's defined in otel?

We should give user the same guidance that we give for attribute names - 

https://github.com/open-telemetry/semantic-conventions/blob/3c16c802e8ae8849ae0cf31eac02c3cabf64e4dd/docs/general/attribute-naming.md?plain=1#L132-L136

We don't attempt to do anything else for attributes, why we would do it for events?

Comment (breedx-splk) [reply docs/attributes-registry/log.md:46]: I think that changing or eliminating `event.name` would be a step backwards at this point. 

I don't think that the existence of 3rd party event APIs that have two identifiers for an event (one numeric, one readable) should influence the design we've already arrived at in otel. Just put the readable id into the `event.name` attribute and put the numeric id into a long attribute named whatever non-semconv thing you want. Why does it need to be more complicated than that....or are we trying to provide formal mapping semantics for every other Event system that might exist out there?

Comment (cijothomas) [reply docs/attributes-registry/log.md:46]: > It almost feels like we should have kept the Name field on LogRecord. I https://github.com/open-telemetry/opentelemetry-specification/issues/3406#issuecomment-2307268466 that I think we need consider adding it back and then it would replace event.name and the proposed log.record.name.

+1 . Irrespective of that, in implementations,  (.NET, Rust), we store name as a top-level field in `LogRecord`, and while exporting, its sent via attributes.

Comment (cijothomas) [reply docs/attributes-registry/log.md:46]: > I think that changing or eliminating event.name would be a step backwards at this point.

That is not part of the proposal, right?


> Just put the readable id into the event.name attribute and put the numeric id into a long attribute named whatever non-semconv thing you want.

This is not just a .NET specific thing, else we could have just used attribute named "dotnet.ilogger.eventid.id"/similar. The goal is to get something language neutral. .NET's ILogger was used as an example.

OTel C++ has the exact same need.
OTel Rust don't yet have the numerical identifier (it just has the human readable name part), but we are trying to get it added as well.

File: docs/attributes-registry/log.md (hunk 5 above)
Comment (cijothomas) [reply docs/attributes-registry/log.md:45]: Resolving this conversation and we can continue in https://github.com/open-telemetry/semantic-conventions/pull/1339#discussion_r1729415678

File: docs/attributes-registry/log.md (hunk 6 above)
Comment (lmolkova) [reply docs/attributes-registry/log.md:46]: I agree, it seems we have 3 different discussions here
1. use `event.name` vs introduce new `log.record.name`
2. do we need first-class data model Name property on logs
3. how to capture id (that appears in several languages and also in windows event log).

Maybe we can focus on the p1 in this thread

Comment (tedsuo) [reply docs/attributes-registry/log.md:46]: Yes back on topic! `log.record.name`!

My suggestion is that we separate `log.record.name` and `log.record.id`. The ID attribute is clear cut and valuable, not to mention a big deal. My request would be to please reduce this PR so that it only contains the ID attribute.  That can go ahead and be approved based on its own merits.

If `log.record.name` really is incompatible with our definition of `event.name`, please make a separate PR for `log.record.name` that clearly presents which issues the additional name field would address. ✌️ 

### on the name field thing
Philosophically I agree that Name should be a field on LogRecord. That would match the way we handle it on Span. But in this particular case, I believe that the Name field actually complicates things. 

This is the problem: we want to support the use case where 3rd party loggers can create events on their own. Those loggers have no concept of a `Name` field, so they would have to use the `event.name` attribute pattern.  Which means that there would have to be code in various places that scans for the `event.name` attribute, and then tries to delete that attribute and move the value over to the Name field.

Because support for 3rd party loggers is critical, there is no escaping the existence of the `event.name` attribute in many parts of our logging pipeline. Which means that the choice is not between "do you want to only have an `event.name` attribute?" and "do you want to only have an `Name` field?", it is actually between "do you want to only have an `event.name` attribute?" or "do you want to have both an `event.name` attribute plus a `Name` field you have to check?"

Long story short: It appears to be the case that a lot of double checking and confusion could be avoided if we just stick with the attribute `event.name`. Solving problems by doing nothing, this is The Way. :)

Comment (cijothomas) [reply docs/attributes-registry/log.md:46]: Based on these discussions/feedback, this is my understanding:
1. There is general agreement that there is no need of introducing `log.record.name` as it is same as the `event.name`. That makes every log done using .NET's ILogger with EventName (_eg: logger.Log(new EventId(10, "LoginFailed"), body, attributes);_) an OTel Event, as it is a `LogRecord` with an attribute "event.name". No need of spec/semantic convention changes, OTel .NET can fix its implementation to store EventName as "event.name".

2. The numerical version of the "event.name", could be named "event.id", call out that "event.id" and "event.name" should have 1:1 relation in a well behaved app. This can be a separate PR.

example:
```
var logger = GetLogger("company.product.service.module");
logger.Log(new EventId(10, "LoginFailed"), body, attributes);
```

The above will produce a LogRecord with scope name set to "company.product.service.module", and attribute containing "event.name" = "LoginFailed", and "event.id" = 10 (along with body and other attributes)

Is this everyone's understanding as well?
